src/stream/src/cache/range_lru.rs

```diff
@@ -0,0 +1,542 @@
+                (false, new_existing_range)
+            } else {
+                panic!("The evicted range {range:?} does not share a bound with the existing range {existing_range:?}");
+            };
+
+            // If from top, delete top part of the range
+            if !from_bottom {
+                // If from top remove the top marker as well.
+                match range.1.clone() {
+                    Included(x) => {
+                        assert!(self.secondary_index.remove(&(x, true)).unwrap().is_none())
+                    }
+                    Excluded(x) => {
+                        assert!(self.secondary_index.remove(&(x, false)).unwrap().is_none())
+                    }
+                    Unbounded => (),
+                }
+            }
+            while let Some((k, v)) = iter.next() {
+                match translate_lower_bound(v.0.clone()) {
+                    Included(x) => {
+                        if from_bottom || iter.peek().is_some() {
+                            self.secondary_index.remove(&x).unwrap(); // TODO: assert the epochs are
+                                                                      // the same?
+                        } else {
+                            // From top and last element, we should simply convert it to `None`
+                            // instead of removing
+                            let old = self.secondary_index.get_mut(&x).unwrap();
+                            if let Some(e) = old {
+                                assert_eq!(*e, *k);
+                            }
+                            *old = None;
+                        }
+                    }
+                    Unbounded => assert_eq!(self.lower_unbounded_epoch.take().unwrap(), *k),
+                    _ => unreachable!(),
+                }
+            }
+            Some((range, new_existing_range))
+        } else {
+            None
+        }
+    }
+
+    /// `access` is called when a range is accessed during an epoch. `access` assumes that
+    /// the ranges provided form a continuous path as a function of the epochs.
+    pub(crate) fn access(&mut self, range: ScalarRange, epoch: u64) {
+        assert_ne!(epoch, 0);
+        assert!(self.last_accessed_epoch < epoch);
+        self.last_accessed_epoch = epoch;
+        let new_existing_range = self
+            .existing_range
+            .take()
+            .map_or(Some(range.clone()), |er| {
+                Some(merge_continuous_ranges(range.clone(), er))
+            });
+        self.existing_range = new_existing_range;
+
+        let mut last_deleted_epoch = None;
+        let mut delete_upper_range = false;
+        let mut remove_keys = vec![];
+        {
+            let encoded_range = (
+                translate_lower_bound(range.0.clone()),
+                translate_upper_bound(range.1.clone()),
+            );
+
+            let mut iter = self.secondary_index.range(encoded_range).peekable();
```

Comment:
Let us also add the `lower_unbounded_epoch` to this iter if `encoded_range.0` ==  `Unbounded`,